Move versioned Apple LLVM targets from `rustc_target` to `rustc_codegen_ssa`

madsmtm opened the pull request:
Fully specified LLVM targets contain the OS version on macOS/iOS/tvOS/watchOS/visionOS, and this version depends on the deployment target environment variables like `MACOSX_DEPLOYMENT_TARGET`, `IPHONEOS_DEPLOYMENT_TARGET` etc.

We would like to move this to later in the compilation pipeline, both because it feels impure to access environment variables when fetching target information, but mostly because we need access to more information from https://github.com/rust-lang/rust/pull/130883 to do https://github.com/rust-lang/rust/issues/118204. See also https://github.com/rust-lang/rust/pull/129342#issuecomment-2335156119 for some discussion.

The first and second commit does the actual refactor, it should be a non-functional change, the third commit adds diagnostics for invalid deployment targets, which are now possible to do because we have access to the session.

Tested with the same commands as in https://github.com/rust-lang/rust/pull/130435.

r? @petrochenkov

madsmtm committed:
Emit diagnostics for incorrect deployment targets

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11095837207/job/30824960447) [(plain)](https://github.com/rust-lang/rust/commit/c4b18fcf9e300a4d7662b0e0f7121f0a8281037c/checks/30824960447/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.002 Building wheels for collected packages: reuse
#13 3.003   Building wheel for reuse (pyproject.toml): started
#13 3.267   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.268   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.268   Stored in directory: /tmp/pip-ephem-wheel-cache-rilrphp5/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.270 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.684 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.685 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.253 Collecting virtualenv
#13 4.253 Collecting virtualenv
#13 4.312   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.530      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 27.8 MB/s eta 0:00:00
#13 4.594 Collecting filelock<4,>=3.12.2
#13 4.602   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.640 Collecting platformdirs<5,>=3.9.1
#13 4.648   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.672 Collecting distlib<1,>=0.3.7
#13 4.690      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 51.6 MB/s eta 0:00:00
#13 4.690      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 51.6 MB/s eta 0:00:00
#13 4.770 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.961 Successfully installed distlib-0.3.8 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 5.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      182208 kB
DirectMap2M:     8206336 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
##[endgroup]
fmt check
fmt: checked 5596 files
tidy check
tidy error: /checkout/compiler/rustc_codegen_ssa/messages.ftl: message `codegen_ssa_apple_deployment_target_too_low` appears before `codegen_ssa_apple_deployment_target_invalid`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.2)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sun Sep 29 20:59:00 UTC 2024
  network time: Sun, 29 Sep 2024 20:59:00 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bjorn3 commented on compiler/rustc_codegen_cranelift/src/lib.rs:
```diff
@@ -259,7 +260,7 @@ impl CodegenBackend for CraneliftCodegenBackend {
 
 fn target_triple(sess: &Session) -> target_lexicon::Triple {
-    match sess.target.llvm_target.parse() {
+    match versioned_llvm_target(sess).parse() {
```
This should probably remain unchanged. Cranelift currently doesn't consume the deployment target and cranelift-object may well expect it to be passed in separately in the future. In general Cranelift barely takes any information from the target triple. Only the cpu architecture and default calling convention for the OS currently. Even the object file format dependent way to lower TLS is passed in using a separate flag.

madsmtm replied:
Yeah, [we talked about this elsewhere](https://github.com/bytecodealliance/wasmtime/issues/8730#issuecomment-2380725262), I did it like this for now so that the commit was completely free of functional changes.

In particular, I think passing an unversioned target triple to `target-lexicon` would result in an error here, as [their current parsing code for `macosx`](https://github.com/bytecodealliance/target-lexicon/blob/7c80d459a9fdd121e9f23feb680c3db13c1baa39/src/targets.rs#L1392-L1421) assumes that the version is always present.

madsmtm replied:
I'll try to whip up a PR to fix this in `target-lexicon`, but that shouldn't block this PR IMO.

madsmtm replied:
Done, see https://github.com/bytecodealliance/target-lexicon/pull/113

bors commented:
:umbrella: The latest upstream changes (presumably #131158) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented on compiler/rustc_codegen_ssa/src/apple.rs:
```diff
@@ -0,0 +1,69 @@
+    if sess.target.is_like_osx {
+        add_version_to_llvm_target(&sess.target.llvm_target, deployment_target(sess)).into()
+    } else {
+        sess.target.llvm_target.clone()
```
The function already returns a `Cow` so you could use a reference here.

petrochenkov commented on compiler/rustc_codegen_ssa/src/apple.rs:
```diff
@@ -0,0 +1,69 @@
+/// Mach-O commands.
+///
+/// Certain optimizations also depend on the deployment target.
+pub fn versioned_llvm_target(sess: &Session) -> Cow<'static, str> {
```
Could you move this function to `rustc_codegen_ssa::back`?
It's used on all platforms, not just apple.

petrochenkov replied:
Ah, and could you move `mod apple;` from the root to there too?

madsmtm replied:
None of the modules in `rustc_codegen_ssa::back` seemed to really semantically match, so I put it in `rustc_codegen_ssa/src/base.rs` instead, feel free to tell me to do something different?

petrochenkov commented on compiler/rustc_codegen_ssa/src/apple.rs:
```diff
@@ -1,30 +1,58 @@
 /// Get the deployment target based on the standard environment variables, or fall back to the
 /// minimum version supported by `rustc`.
 pub fn deployment_target(sess: &Session) -> AppleOSVersion {
+    let os_min = apple_os_minimum_deployment_target(&sess.target.os);
```
Could you move this to the point of use?

petrochenkov commented on compiler/rustc_codegen_ssa/src/apple.rs:
```diff
@@ -1,30 +1,58 @@
+                version.max(min)
+            }
+            Err(error) => {
+                sess.dcx().emit_err(AppleDeploymentTarget::Invalid { env_var, error });
```
Technically this can break builds that worked before by falling back to the minimum version.
Can it happen in realistic scenarios?

madsmtm replied:
> Technically this can break builds that worked before by falling back to the minimum version. Can it happen in realistic scenarios?

Depends on what you mean by "realistic scenarios". I neither know nor can think of a case where setting the variable invalidly would ever make sense, which is why I made this a hard error.

But conceivably, the user could have been setting for example `MACOSX_DEPLOYMENT_TARGET=10_14` in their CI, thinking that it would work, and then never touched it again - but then we _want_ to error, because it won't work ;).

petrochenkov replied:
Ok.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -3037,7 +3038,7 @@ fn add_apple_link_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavo
         };
 
-        let (major, minor, patch) = current_apple_deployment_target(&sess.target);
+        let (major, minor, patch) = deployment_target(sess);
```
```suggestion
        let (major, minor, patch) = apple::deployment_target(sess);
```
Style nit: import `apple`, not the function, this is specifically an Apple deployment target, not some general deployment target.

petrochenkov commented:
Sorry for the delays.
@rustbot author

madsmtm commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -175,97 +175,62 @@ pub fn platform(target: &Target) -> Option<u32> {
-    // When bumping a version in here, remember to update the platform-support
-    // docs too.
+/// Minimum operating system versions currently supported by `rustc`.
+pub fn os_minimum_deployment_target(os: &str) -> OSVersion {
```
Certainly, I was just under the impression that people wanted to keep as much stuff in `rustc_target` as possible

madsmtm replied:
I went and did this for the last remaining `pub` item in `rustc_target` as well in the second commit

madsmtm commented:
> Sorry for the delays.

No problem! Resolved your concerns, so @rustbot ready

madsmtm commented:
I just found out today that the `powerpc-unknown-freebsd` seems to use a (hardcoded) versioned LLVM target as well - should I also move that to `rustc_codegen_ssa`?

petrochenkov commented:
r=me with a couple of visibilities reduced.
@rustbot author

madsmtm committed:
Move Mach-O platform information to rustc_codegen_ssa::back::apple

To align with the general decision to have this sort of information
there instead.

Also use the visionOS values added in newer `object` release.

madsmtm committed:
Move versioned LLVM target creation to rustc_codegen_ssa

The OS version depends on the deployment target environment variables,
the access of which we want to move to later in the compilation pipeline
that has access to more information, for example `env_depinfo`.

madsmtm commented:
@rustbot ready

> > I just found out when going through `cc` code that the `powerpc-unknown-freebsd` seems to use a (hardcoded) versioned LLVM target as well - should I also move that to `rustc_codegen_ssa`?
> 
> This regex `llvm_target: ".*\..*"` finds 2 versioned built-in targets - `aarch64-unknown-solaris2.11` and `powerpc-unknown-freebsd13.0`. Not sure how target versions work for those targets, it's probably better to ask their maintainers, and keep this PR apple-centric.

I've added a FIXME for now.

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 1ef1af1c607fe268b5622041ff586ca603b8ec8d has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 1ef1af1c607fe268b5622041ff586ca603b8ec8d -->
<!-- homu: {"type":"Approved","sha":"1ef1af1c607fe268b5622041ff586ca603b8ec8d","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
